src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+            raise ValueError(
+                "input_ids exceeds maximum length: {} > {}".format(input_ids.size(-1), self.max_positions())
+            )
+        encoder_out = self.encoder(input_ids)
+        prev_output_tokens = self.shift_tokens_left(input_ids, self.config.pad_token_id)
+        dec_features, dec_hidden, dec_attn = self.decoder(prev_output_tokens, encoder_out=encoder_out,)
+        if return_for_head:  # split encoder and decoder outputs nicely
+            return (
+                _filter_out_nones(dec_features, dec_hidden, dec_attn),
+                _filter_out_nones(encoder_out.encoder_out, encoder_out.encoder_states, encoder_out.encoder_attn),
+            )
+        if self.output_hidden_states and self.output_attentions:
+            return (
+                dec_features,
+                dec_hidden,
+                dec_attn,
+                encoder_out.encoder_out,
+                encoder_out.encoder_states,
+                encoder_out.encoder_attn,
+            )
+        elif self.output_hidden_states:
+            return (
+                dec_features,
+                dec_hidden,
+                encoder_out.encoder_out,
+                encoder_out.encoder_states,
+            )
+        elif self.output_attentions:
+            return (dec_features, dec_attn, encoder_out.encoder_out, encoder_out.encoder_attn)
+        else:
+            return (dec_features, encoder_out.encoder_out)
+
+    @staticmethod
+    def shift_tokens_left(input_ids, pad_token_id):
+        """Shift input ids one token to the left"""
+        prev_output_tokens = input_ids.clone()
+        prev_output_tokens[:, 0] = input_ids.gather(
+            1, (input_ids.ne(pad_token_id).sum(dim=1) - 1).unsqueeze(-1),
+        ).squeeze()
+        prev_output_tokens[:, 1:] = input_ids[:, :-1]
+        return prev_output_tokens
+
+    def max_positions(self):
+        """Maximum length supported by the model."""
+        return (self.encoder.max_positions(), self.decoder.max_positions())
+
+
+class BartForMaskedLM(PretrainedBartModel):
+    base_model_prefix = "model"
+
+    def __init__(self, config: BartConfig):
+        super().__init__(config)
+        self.model = BartModel(config)
+
+    def forward(self, *args, **kwargs):  # TODO(SS): better signature when base API decided
+        kwargs["return_for_head"] = True
+        lm_labels = kwargs.pop("lm_labels", None)
+        decoder_outputs, encoder_outputs = self.model(*args, **kwargs)
+        lm_logits = self.output_layer(decoder_outputs[0])
+        decoder_outputs = (lm_logits,) + decoder_outputs[1:]  # Add hidden states and attention if they are here
+
+        if lm_labels is not None:
+            loss_fct = nn.CrossEntropyLoss()
+            masked_lm_loss = loss_fct(lm_logits.view(-1, self.config.vocab_size), lm_labels.view(-1))
+            decoder_outputs = (masked_lm_loss,) + decoder_outputs
+        return decoder_outputs + encoder_outputs
+
+    def output_layer(self, x):
```

Comment:
This is a nice approach.

Though, usually we do weight tying by having a `nn.Linear` layer referenced in the `get_output_embedding` method and the `init_weights` method is in charge of tying the weights (see the base class `PretrainedModel`). See an example in T5: https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_t5.py#L848-L859

The reason we do this (slightly more complex) setup is that it allows to deactivate weights tying. That's debatable though because weights are always tied today. Overall I would say:
- it's best to have the same process in every models, so probably good to align this model's behavior with the other models
- torchscript doesn't like our way of doing weights tying, maybe this could be a solution cc @LysandreJik
- if we decide to go that way here, we should update the other models so that every model uses `F.linear`.

Reply:
aligning with other models